boost::container_hash fails to build with clang/libc++ 15.0.0

I upgraded my workstation to Fedora 37 which has the 15.0.0 release of clang/libc++, and I'm now getting compile errors with the latest commit of rules_boost, with output like this:
```
In file included from bzlup/opts.cc:7:
In file included from external/boost/boost/program_options.hpp:15:
In file included from external/boost/boost/program_options/options_description.hpp:13:
In file included from external/boost/boost/program_options/value_semantic.hpp:12:
In file included from external/boost/boost/any.hpp:20:
In file included from external/boost/boost/type_index.hpp:29:
In file included from external/boost/boost/type_index/stl_type_index.hpp:47:
external/boost/boost/container_hash/hash.hpp:132:33: error: no template named 'unary_function' in namespace 'std'; did you mean '__unary_function'?
        struct hash_base : std::unary_function<T, std::size_t> {};
                           ~~~~~^
/usr/bin/../include/c++/v1/__functional/unary_function.h:46:1: note: '__unary_function' declared here
using __unary_function = __unary_function_keep_layout_base<_Arg, _Result>;
^
```

I haven't tested yet but from https://www.boost.org/doc/libs/1_80_0/libs/config/doc/html/boost_config/boost_macro_reference.html and looking at the header file it appears that this can be fixed by defining `BOOST_NO_CXX98_FUNCTION_BASE`. However defining this might cause problems for people using older compilers?

--- a/BUILD.boost
+++ b/BUILD.boost
@@ -628,6 +628,7 @@ boost_library(
 
 boost_library(
     name = "container_hash",
+    defines = ["BOOST_NO_CXX98_FUNCTION_BASE"],
     deps = [
         ":assert",
         ":config",

